WIP Convert Filebeat kibana.log to ECS

webmat opened the pull request:
**Caveats**

- [ ] Current Kibana module populates `service.name` with an array of one service name. Not sure what the purpose is for this? Should we change back to a single value?
- [ ] Should we ask the Kibana team to migrate their log format to ECS?
- [ ] This module populates `http.response.elapsed_time` which is not in ECS.
- [ ] This module populates `http.response.content_length` which is not in ECS.

**Renames**

- kibana.log.meta.req.headers.referer => http.request.referrer
- kibana.log.meta.req.headers.user-agent => user_agent.original
- kibana.log.meta.req.remoteAddress => source.ip
- kibana.log.meta.req.url => url.original

**Duplicate fields**

They have been removed, but will be aliased

- kibana.log.meta.req.referer => http.request.referrer
- kibana.log.meta.statusCode => http.response.status_code (ECS field was already populated)
- kibana.log.meta.method => http.request.method

**TODO**

- [x] Force `http.request.method` to uppercase `http.request.method`
- [ ] split host into destination domain and port?
- [ ] Populate event.duration based on http.response.elapsed_time?
- [ ] Alias renamed fields to their ECS counterpart
- [ ] Alias duplicate fields that have been removed to their ECS counterpart
- [ ] Document field migrations in ecs-migration.yml
- [ ] Changelog
- [ ] Perform user agent parsing?
- [ ] Perform geoip on source.ip?

ruflin commented:
Caveats:
* service.name: I remember it has to do on how we add service.name. From a lucene perspective it's the same so not need to change. But there is an issue. This should become `service.type` as `service.name` can be a user configured name.
* We should open an issue in the KB repo to start a discussion around ECS logs
* Missing fields: Let's open track this during the discussion related to http fields

webmat commented:
@ruflin Apart from one thing that I wonder about (`kibana.log.tags` >> `tags`?), this should be pretty much ready to go.

webmat commented:
jenkins, test this

ruflin commented on filebeat/module/kibana/log/test/test.log-expected.json:
```diff
@@ -17,21 +18,18 @@
-        "kibana.log.meta.req.remoteAddress": "127.0.0.1", 
-        "kibana.log.meta.req.url": "/ui/fonts/open_sans/open_sans_v15_latin_600.woff2", 
         "kibana.log.meta.req.userAgent": "127.0.0.1", 
-        "kibana.log.meta.statusCode": 304, 
```
Where did this one go? I think a few fields are missing now?

ruflin replied:
Deleting as I saw this was removed based on the PR description.

ruflin commented:
@webmat I'm good with keep kibana.logs.tags as I think these are tags specific to kibana.

webmat commented:
@ruflin This is correct. Some fields were present twice in the output log, under different names. Now they are present only once, in line with ECS, and all the old names are aliases to the ECS fields.

webmat commented:
@ruflin One more caveat I'd like your opinion on: `service.name`, we leave as an array?

ruflin commented:
@webmat Yes, for the array. It does not make a difference. Can explain an other time why it's like.

After we done all the modules, will go through an rename service.name to service.type anyways.

webmat commented:
@ruflin You had approved this before the holidays, I tweaked one small detail. Instead of renaming `source.address` to `source.ip`, I copy it (keeping `source.address`).

Can I get one last look, please? Thanks

webmat commented:
Ping @elastic/stack-monitoring for awareness. If this works for you as well, I'd like to merge on Tuesday at the latest.

ycombinator commented on filebeat/module/kibana/log/ingest/pipeline.json:
```diff
@@ -53,6 +53,14 @@
+                "lang": "painless",
+                "source": "ctx.event.duration = Math.round(ctx.kibana.log.meta.res.responseTime * params.scale)",
+                "params": { "scale": 1000000 },
+                "if": "ctx.kibana.log.containsKey('meta') && ctx.kibana.log.meta.containsKey('res') && ctx.kibana.log.meta.res.containsKey('responseTime')"
```
Nit: You could probably shorten this to something like: `ctx.kibana.log.meta?.res?.responseTime != null`.

webmat replied:
Oh man, Painless has that?  Thanks for pointing out ❤️

ycombinator replied:
Yep, it was added at some point: https://www.elastic.co/guide/en/elasticsearch/painless/current/painless-operators-reference.html#null-safe-operator

ruflin replied:
@ycombinator Do you know which version this was added? Asking for compatibility reasons.

ycombinator replied:
5.1.1: https://github.com/elastic/elasticsearch/pull/21239

ruflin replied:
Ah, then we are good :-)

ruflin commented:
jenkins, test this

webmat commented:
Rebased and updated the pipeline to use the null-safe operators. Thanks @ycombinator for pointing those out, this will make my life much easier in places like this experimental PR: #9905 😆 

Tests were green across the board. Will merge tomorrow AM if it's still the case.
